### PR TITLE
added raincloudPlots

### PR DESCRIPTION
**Pull request by vincentott**
Requires the following JASP nightly:
macOS arm64 from 06:33 21-03-2024

Importantly, I did not run the unit tests in:
test-descriptives.R
and
test-descriptives-verified.R

as I encountered some installation errors for:
jaspTools::testAll()
with respect to the two files.
Thus, I only ran:
jaspTools::testAnalysis("raincloudPlots")

Please test, whether
jaspTools::testAll()
works as intended!

### CODE REVIEW

File: R/raincloudPlots.R
Comment (vandenman) [inline R/raincloudPlots.R:477]: Is this the only use case for `dplyr`? If yes, I would drop it as a dependency and do this is in a different way.

File: R/raincloudPlots.R
Comment (vandenman) [inline R/raincloudPlots.R:458]: This looks a bit sus, why do we need this? Aren't there other ways to determine this without building the plot?

File: inst/qml/raincloudPlots.qml
Comment (vandenman) [inline inst/qml/raincloudPlots.qml:196]:
```suggestion
			checked: 	!(meanInterval.checked || meanIntervalCustom.checked)
```

File: inst/qml/raincloudPlots.qml
Comment (vandenman) [inline inst/qml/raincloudPlots.qml:209]: @JorisGoosen I don't think this is easier tot translate, no?

File: R/raincloudPlotsWrapper.R
Comment (vandenman) [inline R/raincloudPlotsWrapper.R:36]: Isn't this is a bit of a weird default value?

File: R/raincloudPlots.R
Comment (vandenman) [inline R/raincloudPlots.R:183]: I don't think the empty `else` statements are necessary, nor are the comments `} End xxx`.

File: R/raincloudPlots.R
Comment (vandenman) [inline R/raincloudPlots.R:800]:
```suggestion
  tableInProgress$addColumnInfo(name = "primaryFactor", title = gettext("Primary Factor"), type = "string")
```
same for all other titles of the table columns.

File: R/raincloudPlots.R
Comment (vandenman) [inline R/raincloudPlots.R:901]: please rewrite using `gettextf`, this cannot be translated.

File: R/raincloudPlots.R
Comment (vandenman) [inline R/raincloudPlots.R:906]: Often, it is better to write one series of if statements _with_ code duplication. For example:

```r
footnote <- NULL
if (options$meanIntervalCustom) {
  if (dataInfo$numberOfExclusions == 0 && 
	  !(options$meanIntervalCustom || !(options$observationId == "" && options$meanInterval && options$meanIntervalOption == "ci" && options$meanCiAssumption)))
    gettextf("<i>N</i><sub>Total</sub> = %s.", dataInfo$sampleSize)
  else if (dataInfo$numberOfExclusions == 0) {
    if (options$meanIntervalCustom) {
      gettextf("<i>N</i><sub>Total</sub> = %s. Interval around mean is custom.", dataInfo$sampleSize)
    } else {
      # etc.
    }
  }
```
While the code becomes objectively worse, the individual strings become easier to translate. Now, in this case, I think the sentences are distinct enough that it doesn't really matter so whether you want to do this is up to you.

File: R/raincloudPlots.R
Comment (vandenman) [inline R/raincloudPlots.R:622]: should this be an error the user sees?

File: R/raincloudPlots.R
Comment (vandenman) [inline R/raincloudPlots.R:727]:
```suggestion
  jitter <- if (options$jitter) {
    gettextf("Points are slightly jittered from their true values.")
  }
```
because `a <- if (FALSE) 1` implies that `a` is `NULL`.

File: R/raincloudPlotsWrapper.R
Comment (vandenman) [inline R/raincloudPlotsWrapper.R:35]: I don't think this default argument makes sense?

**Review by vandenman — CHANGES_REQUESTED** [review]
Can you regenerate the failing test figures? When I do that then locally the tests pass.

File: R/raincloudPlots.R
Comment (vincentott) [reply R/raincloudPlots.R:616]: no, that was a left-over from debugging. i removed the else statement altogether.

File: R/raincloudPlots.R
Comment (vincentott) [reply R/raincloudPlots.R:452]: This was my rationale:

Goal:
points, vioOutline, boxOutline, mean, meanLines should have the option to be shown in the group color according to the colorPalette.

Required info:
group color(s) according to the colorPalette

Problem:
Depending on the number of groups,
`jaspGraphs::scale_JASPfill_discrete()` and `jaspGraphs::scale_JASPcolor_discrete()`
as applied to a ggplot
will yield different hexcodes. If you have two (vs. three) groups, then you get two (vs. three) hexcodes.
**However**, the first two hexcodes did not always match when I tried this out (maybe this is also specific to some of the colorPalettes?).

Solution:
First apply the colorPalette (section Palettes in `.rainFillPlot()`), then extract the assigned hexcodes through `ggplot_build()` (see `.rainInfoFactorCombinations()`, Preparation section in `.rainFillPlot()`).

Because not every plot will have a fill (requires secondaryFactor or colorAnyway),
it will not always be possible to extract the hexcodes. This is why I used `tryCatch()`.
I guess I could do an if-else statement instead of `tryCatch()` (might be a bit cleaner?).
But as I remember it, I found no work-around and ended up having to `ggplot_build()`.
Because only then you have access to the group colors from the colorPalette in the way ggplot2 decided to assign them to the groups.

Please let me know, what you think! 😄

File: R/raincloudPlots.R
Comment (vincentott) [reply R/raincloudPlots.R]: Because it is still possible to translate the strings, I would like to keep the code tidier.
I made minor changes so that the footnote works even if there are no exclusions or meanInterval,
while translators do not have to wrangle with random spaces at the beginning of a string.

File: inst/qml/raincloudPlots.qml
Comment (vincentott) [reply inst/qml/raincloudPlots.qml:204]: I do not fully understand this comment and will wait for Joris´ response.

File: R/raincloudPlotsWrapper.R
Comment (vincentott) [reply R/raincloudPlotsWrapper.R]: yes, this wrapper got created from penguins.jasp
in fact, only the three value arguments are relevant for the module.
i simplified the customizationTable default in the wrapper and the unit tests.

File: R/raincloudPlotsWrapper.R
Comment (vincentott) [reply R/raincloudPlotsWrapper.R]: yes, this wrapper got created from penguins.jasp
in fact, only the three value arguments are relevant for the module.
i simplified the customizationTable default in the wrapper and the unit tests.

File: inst/qml/raincloudPlots.qml
Comment (JorisGoosen) [inline inst/qml/raincloudPlots.qml:250]:
```suggestion
			info: 		qsTr(
							"Whether or not the points of a cloud should be shown. If un-checked, opacity is set to 0.<br>\n"
							"If you have many, many points, it can be helpful to hide them.\n"

							"<br><br>\n"

							"<h3>Element Settings</h3>\n"  // All of this here, because infoLabel{} for DoubleField{} does not work.

							"<h4>Nudge</h4>\n"
							"The nudge for violin, box, and point determines how far the elements are nudged from the center (axis tick).<br>\n"
							"By default, the box is in the center (nudge = 0) and the violin is nudged to the right. \n"
							"The points are nudged to the left of the box.<br>\n"
							"With a custom orientation (see Advanced section), the points get centered (nudge fixed to 0). Violin and box are nudged depending on orientation: left or right.\n"

							"<h4>Height, Width, and Spread</h4>\n"
							"These settings determine the respective properties of the elements:<br><br>\n"
							"How high should the violin peak? (Does not change the proportions of the probability density.)<br>\n"
							"With a custom orientation to the left (see Advanced Section), it can be \n"
							"helpful to decrease the height if the violin overlaps with axis ticks.<br><br>\n"
							"How wide should the box be?<br><br>\n"
							"How wide should the points spread? I.e., how jittered should they be along the x-axis (y-axis if horizontal plot?.<br>\n"
							"Note, that this still correctly shows the values of the points on the dependent variable axis.<br>\n"
							"(If you would also like y-jitter see the 'Jitter' option for points.)\n"

							"<h4>Box Padding</h4>\n"
							"With a secondary factor, there will be a box for each factor level. With the padding, you can change the spacing between theses boxes.<br>\n"
							"Increasing the padding will - visually - decrease the box width. If you want to keep box width constant, also increase the box width.\n"

							"<h4>Point Size</h4>\n"
							"We recommend a greater size, the fewer points you have.\n"

							"<h4>Opacity</h4>\n"
							"Increase this to _decrease_ the transparency of the respective element.\n"

							"<h4>Outline and Outline Width</h4>\n"
							"Would you like the outline to match the 'Color palette', be 'black', or have 'none' at all?<br>\n"
							"The width determines how thick the outline is.\n"

							"<h4>Violin Smoothing</h4>\n"
							"This percentage determines the smoothness of the probability density.<br>\n"
							"The lower, the stronger it is influenced by the presence/absence of individual points.\n"

							"<h4>Point Jitter</h4>\n"
							"Use this with care!<br>\n"
							"It slightly jitters the points along the dependent variable axis. This means that the position of the points no longer matches the values of the observations.<br>\n"
							"If you want to have less overlap between the points while keeping an accurate visualization, increases the point spread.<br>\n"
							"Using point jitter can be useful in some cases. For example, if you are working with likert data.\n"

							"<h4>ID Lines Width</h4>\n"
```

Comment (JorisGoosen) [reply inst/qml/raincloudPlots.qml:250]: I dont think you need the `+` everywhere.

Comment (vincentott) [reply inst/qml/raincloudPlots.qml:250]: I just tried dropping the `+` but then the GUI won´t load. So I will leave it in, okay?

File: inst/qml/raincloudPlots.qml
Comment (vincentott) [reply inst/qml/raincloudPlots.qml:204]: Would it be okay to leave it as is and to integrate the rainclouds into one of the upcoming nightlies?

**Comment by vincentott** [comment]
@vandenman Do you know why the checks are suddenly failing?

**Comment by shun2wang** [comment]
>@vandenman Do you know why the checks are suddenly failing?

for i18nCheck failing I will fix it at:https://github.com/jasp-stats/jasp-actions/pull/38

and unittest on macOS failing not related the analysis so you may re-run it again.

**Comment by vincentott** [comment]
> btw: could you also update the analysis help docs in `.md` file? with thanks.

As I understand, there is no need because the help is generated by info{} in the .qml?
So, the help is already there, right? At least I can read it when I click on it in JASP.

**Comment by shun2wang** [comment]
ah, yes I see. nevermind.

**Comment by JorisGoosen** [comment]
I just wanted to look at the raincloud plots to see about formatting those info fields.

But I get this:
image

**Comment by JorisGoosen** [comment]
> How is ContBinom allowed in the Dependent Variable Box, is what I am wondering

sounds like an easy fix to me

File: inst/qml/raincloudPlots.qml
Comment (vincentott) [reply inst/qml/raincloudPlots.qml]: Ah, I originally only removed the `+`s but not the quotes at the end of each line!
Your commit works for me now, however I was not able to change the indentation without also getting a messy layout in the documentation (see pictures).
Is there a way to fix this? Otherwise I would prefer `+` and `"` at the end of each line so that it looks tidier (at least in my vscode).

Screenshot 2024-04-05 at 19 24 23
Screenshot 2024-04-05 at 19 24 40

**Comment by vincentott** [comment]
> How is ContBinom allowed in the Dependent Variable Box, is what I am wondering

I changed it so that only variables of type scale are allowed as dependentVariables.

File: inst/qml/raincloudPlots.qml
Comment (JorisGoosen) [reply inst/qml/raincloudPlots.qml]: Just remove the leading spaces...

Comment (JorisGoosen) [reply inst/qml/raincloudPlots.qml]: `shift+tab` is your friend ;)

Comment (vincentott) [reply inst/qml/raincloudPlots.qml]: So you mean like this?
Screenshot 2024-04-07 at 20 33 49

with documentation like:

Screenshot 2024-04-07 at 20 34 25

I do not like it, as I said above:
Without the leading spaces, the code is not indented. However, I would like it indented.
> Is there a way to fix this?
> Otherwise I would prefer + and " at the end of each line so that it looks tidier

Comment (JorisGoosen) [reply inst/qml/raincloudPlots.qml]: Well, I presonally like being able to just easily add and edit lines in a codebase more important than correct indentation. You know, if someone wants to add a paragraph to your `info` they have to do so in strings `\n"+` etc.
With my way you just type...

So this is a convenience thing because Im all for beautiful indentation and spacing in general!

However, im not very likely to be the one to have to work on this so as long as it ends up in the translation files correctly I dont really mind which one you use ;)

Comment (vincentott) [reply inst/qml/raincloudPlots.qml]: Ah okay, does it end up in the translation files correctly either way?
How do we check?

If it does, may I revert your commit do go back to `"` and `+`?

Comment (JorisGoosen) [reply inst/qml/raincloudPlots.qml]: If it ends up in the *.po files fine of course you can revert it

Comment (vincentott) [reply inst/qml/raincloudPlots.qml:250]: Alright, so can we now pull the rainclouds into the next nightly?
Or is there anything else?

Comment (shun2wang) [reply inst/qml/raincloudPlots.qml:250]: That's fine with `+` to concatenate strings and just we need to pay attention to make sure it is in a complete `qsTr()` call.will generate a complete block. in .po file will looks like:
```po
#: nst/qml/raincloudPlots.qml:607
msgctxt "raincloudPlots|"
msgid ""
"Whether to also show an interval around the mean.<br>This hides the box, but "
"if you really want, you can go up and check 'Show box' again.<br><br>You can "
"choose between ±1 standard deviation or a confidence interval.<br>The "
"confidence interval is computed independently for each group. For this, the "
"corresponding checkbox needs to be checked, to acknowledge that.<br>This "
"computation means that any factors are treated as between-factors.<br>If "
"your data does not meet this assumption, but you still want to show "
"intervals around the means, you can specify custom intervals (see "
"below).<br>One example where this assumption does not hold is with ID input. "
"This is why ID input disables the computation of confidence "
"intervals.<br><h4>Method</h4>How should the confidence interval be "
"computed?<br>'Normal model' uses a standard error defined as: "
"_SD_<sub>group</sub> / square root of _N_<sub>group</sub><br>'T model' "
"yields results identical to a one-sample t-test.<br>For 'Bootstrap' you can "
"further specify the number of samples and set a seed for reproducible "
"results."
msgstr ""
```

Comment (shun2wang) [reply inst/qml/raincloudPlots.qml:250]: >Without the leading spaces, the code is not indented. However, I would like it indented.

for indent things maybe we should just add the style rule in .css for help docs. so developer shouldn't add indent manually? what do you think @JorisGoosen

Comment (JorisGoosen) [reply inst/qml/raincloudPlots.qml:250]: Well, the indent is generated by markdown in any case.

Actually, that is a good point. @vincentott is there a reason you are typing html there and not just markdown?
Because I think it is ok as the html will not be stripped by markdown.
That already handles all the indents anyway.

Me and vincent were just talking about the indent in the qml code

**Comment by JorisGoosen** [comment]
@FBartos or @JohnnyDoorn can you review the latest changes by Vincent and see if he addressed the issues noted by Don?
Because Im sure hes got better things to do right now

**Comment by FBartos** [comment]
I will review this. @vincentott , are those all the changes, or are you planning to push more?

**Comment by vincentott** [comment]
> I will review this. @vincentott , are those all the changes, or are you planning to push more?

I am done! You can start your review, thank you very much! 🙂

File: R/raincloudPlots.R
Comment (FBartos) [inline R/raincloudPlots.R:47]: options (and ideally all lists) should be accessed via [[""]] notation rather than "$"
i.e., ```options[["primaryFactor"]]```

please change throughout the analysis
(also see R guide)

File: R/raincloudPlots.R
Comment (FBartos) [reply R/raincloudPlots.R]: there are still empty else statements here, why?

File: R/raincloudPlots.R
Comment (FBartos) [inline R/raincloudPlots.R:334]: place on separate line

File: R/raincloudPlots.R
Comment (FBartos) [inline R/raincloudPlots.R:347]: there is a shortcut for
```x <- if(a == b) TRUE else FALSE```
it is simply
```x <- a == b```

File: R/raincloudPlots.R
Comment (FBartos) [inline R/raincloudPlots.R:561]: again, should be
```showVioGuide <- options[["secondaryFactor"]] == ""```

File: inst/Description.qml
Comment (FBartos) [inline inst/Description.qml:9]: not sure how much is this comment relevant

**Review by FBartos — CHANGES_REQUESTED** [review]

File: R/raincloudPlots.R
Comment (FBartos) [reply R/raincloudPlots.R]: it indeed looks sus and I'm not sure I really comprehend the explanation why we need tryCatch -- I would reserve that function for actual error catching and use if else for all other purposes

File: R/raincloudPlots.R
Comment (FBartos) [reply R/raincloudPlots.R]: I still see this issue on some other places

**Comment by FBartos** [comment]
+ some additional comments in replies to Don's comments

File: inst/qml/raincloudPlots.qml
Comment (vincentott) [reply inst/qml/raincloudPlots.qml:250]: Once I started using html/css in the .R file for plot and table footnotes, I also kept using it in the .qml file.
So, may I leave it as is?

Comment (JorisGoosen) [reply inst/qml/raincloudPlots.qml:250]: yeah you can

File: R/raincloudPlots.R
Comment (vincentott) [reply R/raincloudPlots.R]: I already fixed this with the help of Don, but did not document it here further.
Now, we simply extract as many hexcodes as there are levels in the factor, without building the plot:

Screenshot 2024-04-15 at 16 16 53

File: R/raincloudPlots.R
Comment (vincentott) [reply R/raincloudPlots.R]: Dropped it where I found it. All good?

File: R/raincloudPlots.R
Comment (vincentott) [reply R/raincloudPlots.R:47]: I would like to keep it as is for three reasons:

First, the JASP r-analyses-guide
https://github.com/jasp-stats/jasp-desktop/blob/development/Docs/development/r-analyses-guide.md#step-3---reading-the-dataset
also suggests the $ operator:
Screenshot 2024-04-15 at 16 31 35

Second, Don never brought it up, so I thought it was fine as is.

Third, I think the code is tidier when using a single `$` compared to double brackets + quotes `[[""]]`.
(Unless in those cases where I am accessing the containers.)

Why do you suggest the change?

File: R/raincloudPlots.R
Comment (vincentott) [reply R/raincloudPlots.R:334]: The JASP r-style-guide allows single-line if-else statements:
https://github.com/jasp-stats/jasp-desktop/blob/development/Docs/development/r-style-guide.md
Screenshot 2024-04-15 at 16 41 17

File: R/raincloudPlots.R
Comment (vincentott) [reply R/raincloudPlots.R]: thanks! much tidier!

File: R/raincloudPlots.R
Comment (vincentott) [reply R/raincloudPlots.R]: thanks!

File: inst/Description.qml
Comment (vincentott) [reply inst/Description.qml]: yes, I removed it!

**Comment by vincentott** [comment]
@FBartos just so that you know what is going on:

@boutinb helped me resolve some merge conflicts as he recently added flexplots and some of the timeSeries to jaspDescriptives.
When you now fetch this raincloudPlots branch for your second review,
there will be the flexplot and timeSeries stuff, but all should work fine.

File: R/raincloudPlots.R
Comment (FBartos) [reply R/raincloudPlots.R:334]: ok, good (I tagged this as it looked quite long)

File: R/raincloudPlots.R
Comment (vincentott) [reply R/raincloudPlots.R]: I implemented it for the options.
Thus, now everything should be resolved and the rainclouds can go into the next nightly, right?

**Review by FBartos — APPROVED** [review]
yes, looks good

**Comment by JohnnyDoorn** [comment]
I just played around with it a bit and really awesome stuff @vincentott! Both in terms of GUI and functionality
There was one thing I did not entirely understand was how the "box" element and the "interval around mean" option interact with each other. One toggles the other, but not vice versa. I also did not understand why I cannot have a confidence interval when (I think) there is a continuous "ID" variable (see attached jasp file, change extension to .jasp).
[debugRaincloud.zip](https://github.com/jasp-stats/jaspDescriptives/files/15010992/debugRaincloud.zip)

**Comment by vincentott** [comment]
@JohnnyDoorn

> I just played around with it a bit and really awesome stuff @vincentott! Both in terms of GUI and functionality

Thank you!

> There was one thing I did not entirely understand was how the "box" element and the "interval around mean" option interact with each other. One toggles the other, but not vice versa (so i can still have a boxplot with "double whiskers").

The underlying idea was: Box is shown per default. The user can also add the mean and it will be placed within the box.
So, I would not like the box to toggle the mean interval. Otherwise the user would first have to remove the box before adding the interval; I think it is smoother if clicking the interval just hides the box.
Do you think its better if it is impossible to re-show the box, as long as the meanInterval is activated? (So that there are no double whiskers?). Should I implement it like that?

> I also did not understand why I cannot have a confidence interval when (I think) there is a continuous "ID" variable (see attached jasp file, change extension to .jasp). [debugRaincloud.zip](https://github.com/jasp-stats/jaspDescriptives/files/15010992/debugRaincloud.zip)

I am sorry, I cant open the the file because the rainclouds are not part of a nightly yet. Could you post one or two screenshots?
In general: It was fairly easy to implement CIs that are computed independently for each group (see checkbox right above ciWidth field). However, with an ID, the groups are not independent anymore and thus the CIs must be computed differently; but I had no time to implement that, too. However, user may still compute them elsewhere and specify them as customIntervals.

**Comment by JorisGoosen** [comment]
> I am sorry, I cant open the the file because the rainclouds are not part of a nightly yet.

Uhm, if you install your dev module it will work

**Comment by vincentott** [comment]
> I also did not understand why I cannot have a confidence interval when (I think) there is a continuous "ID" variable (see attached jasp file, change extension to .jasp). [debugRaincloud.zip](https://github.com/jasp-stats/jaspDescriptives/files/15010992/debugRaincloud.zip)

I was now able to look at it (thanks @JorisGoosen). Your file did not have an ID assigned, but I used "V1" as ID and then - as intended - the confidence interval option because disabled (see my previous message).

**Review by RensDofferhoff — APPROVED** [review]

**Comment by JohnnyDoorn** [comment]
So I'm a bit late, with the code-stop and all, and I think it can still be merged, but I ran into some issues with an Andy Field data set: 
[puppiestest2.zip](https://github.com/jasp-stats/jaspDescriptives/files/15067769/puppiestest2.zip)
(change ext)
and when I toggle "show box" (which i guess causes the issue?), there's an error in the violin. Also the descriptives table seems to have wrong N (2/3/1 instead of 5/5/5)? 
image

**Comment by vincentott** [comment]
@JohnnyDoorn both debugRaincloud.zip and puppiestest2.zip do not work for me anymore.
Could it be that they need to be created with the newest version of this branch?
Is your clone up to date with all the commits that happened since?

I suspect that the problem are spacebars in the names of the factor levels.
Could you maybe send the dataset as a .csv? I cannot find it in the JASP data library.

**Comment by JorisGoosen** [comment]
> Mac M2 intel

This is not a thing btw

**Comment by JohnnyDoorn** [comment]
haha oops, corrected it now

**Comment by JorisGoosen** [comment]
image

I also get this...

File: R/raincloudPlots.R
Comment (vincentott) [reply R/raincloudPlots.R:491]: After your comment, I had dropped this line and thus the dplyr dependency.
Instead I crafted multiple lines of base R as a work-around.
However, we later noticed that plotting would crash when there were factor levels with spaces in their names: e.g. factor hairColor with two levels; "dark brunette" and "light blonde".

Then, I tracked the problem down to my base R work-around.
I tried to fix it with just base R, but then decided that using this simple dplyr line was the best solution.
Because it is tidy and did not cost me much more time to implement.

I asked @boutinb if that is fine and he said its okay.

**Pull request merged**
